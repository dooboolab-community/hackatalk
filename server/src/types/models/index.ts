export * from './AuthPayload';
export * from './Post';
export * from './User';
export * from './Scalar';
