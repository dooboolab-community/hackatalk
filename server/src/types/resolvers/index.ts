export * from './User';
export * from './Post';
