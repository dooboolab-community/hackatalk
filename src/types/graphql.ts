export type Maybe<T> = T | null;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
  DateTime: any;
  Upload: any;
};

export type AuthPayload = {
  __typename?: 'AuthPayload';
  token: Scalars['String'];
  user: User;
};

export enum AuthType {
  Email = 'email',
  Facebook = 'facebook',
  Google = 'google',
  Apple = 'apple'
}

export type Channel = {
  __typename?: 'Channel';
  id: Scalars['ID'];
  type?: Maybe<ChannelType>;
  name?: Maybe<Scalars['String']>;
  messages?: Maybe<Array<Maybe<Message>>>;
  memberships?: Maybe<Array<Maybe<Membership>>>;
  myMembership?: Maybe<Membership>;
  createdAt: Scalars['DateTime'];
  updatedAt: Scalars['DateTime'];
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export type ChannelInput = {
  type?: Maybe<ChannelType>;
  name?: Maybe<Scalars['String']>;
  friendIds: Array<Scalars['String']>;
};

export enum ChannelType {
  Private = 'PRIVATE',
  Public = 'PUBLIC'
}



export type File = {
  __typename?: 'File';
  filename: Scalars['String'];
  mimetype: Scalars['String'];
  encoding: Scalars['String'];
};

export type Friend = {
  __typename?: 'Friend';
  id: Scalars['ID'];
  user?: Maybe<User>;
  friend?: Maybe<User>;
  createdAt?: Maybe<Scalars['DateTime']>;
  updatedAt?: Maybe<Scalars['DateTime']>;
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export type FriendPayload = {
  __typename?: 'FriendPayload';
  user: User;
  added?: Maybe<Scalars['Boolean']>;
  deleted?: Maybe<Scalars['Int']>;
};

export type FriendSub = {
  __typename?: 'FriendSub';
  user?: Maybe<User>;
  action?: Maybe<FriendSubAction>;
};

export enum FriendSubAction {
  Added = 'ADDED',
  Updated = 'UPDATED',
  Deleted = 'DELETED'
}

export type Gallery = {
  __typename?: 'Gallery';
  id: Scalars['ID'];
  photoURL: Scalars['String'];
  createdAt: Scalars['DateTime'];
  updatedAt: Scalars['DateTime'];
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export enum Gender {
  Male = 'MALE',
  Female = 'FEMALE'
}

export type Membership = {
  __typename?: 'Membership';
  id: Scalars['ID'];
  channel?: Maybe<Channel>;
  user?: Maybe<User>;
  type?: Maybe<MemberType>;
  userAlert?: Maybe<Scalars['Boolean']>;
  userMode?: Maybe<UserModeType>;
  createdAt?: Maybe<Scalars['DateTime']>;
  updatedAt?: Maybe<Scalars['DateTime']>;
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export enum MemberType {
  Owner = 'OWNER',
  Member = 'MEMBER'
}

export type Message = {
  __typename?: 'Message';
  id: Scalars['String'];
  channel?: Maybe<Channel>;
  sender?: Maybe<User>;
  type?: Maybe<Scalars['String']>;
  text?: Maybe<Scalars['String']>;
  picture?: Maybe<Array<Maybe<Photo>>>;
  filePath?: Maybe<Scalars['String']>;
  replies?: Maybe<Array<Maybe<Reply>>>;
  reactions?: Maybe<Array<Maybe<Reaction>>>;
  createdAt?: Maybe<Scalars['DateTime']>;
  updatedAt?: Maybe<Scalars['DateTime']>;
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export type MessagePayload = {
  __typename?: 'MessagePayload';
  channelId: Scalars['String'];
  message?: Maybe<Message>;
};

export type Mutation = {
  __typename?: 'Mutation';
  signInEmail: AuthPayload;
  signInWithSocialAccount: AuthPayload;
  signInWithFacebook: AuthPayload;
  signUp: AuthPayload;
  findPassword?: Maybe<Scalars['Boolean']>;
  sendVerification?: Maybe<Scalars['Boolean']>;
  addNotificationToken?: Maybe<Notification>;
  removeNotificationToken?: Maybe<Scalars['Int']>;
  updateProfile?: Maybe<User>;
  addFriend?: Maybe<FriendPayload>;
  deleteFriend?: Maybe<FriendPayload>;
  /** `friendIds` in Channel should exclude userid. */
  createChannel?: Maybe<Channel>;
  updateChannel?: Maybe<Scalars['Int']>;
  deleteChannel?: Maybe<Scalars['Int']>;
  /**
   * Create message and return channelId when meessage has successfully sent.
   * Do not pass current userId inside `users`.
   */
  createMessage?: Maybe<MessagePayload>;
  setOnlineStatus?: Maybe<Scalars['Int']>;
  changeEmailPassword?: Maybe<Scalars['Boolean']>;
  createGallery?: Maybe<Gallery>;
  updateGallery?: Maybe<Scalars['Int']>;
  deleteGallery?: Maybe<Scalars['Int']>;
  singleUpload: Scalars['String'];
  createReaction?: Maybe<Reaction>;
  deleteReaction?: Maybe<Scalars['Int']>;
};


export type MutationSignInEmailArgs = {
  email: Scalars['String'];
  password: Scalars['String'];
};


export type MutationSignInWithSocialAccountArgs = {
  socialUser: SocialUserInput;
};


export type MutationSignInWithFacebookArgs = {
  accessToken: Scalars['String'];
};


export type MutationSignUpArgs = {
  user: UserInput;
};


export type MutationFindPasswordArgs = {
  email: Scalars['String'];
};


export type MutationSendVerificationArgs = {
  email: Scalars['String'];
};


export type MutationAddNotificationTokenArgs = {
  notification: NotificationCreateInput;
};


export type MutationRemoveNotificationTokenArgs = {
  token: Scalars['String'];
};


export type MutationUpdateProfileArgs = {
  user: UserProfileInput;
};


export type MutationAddFriendArgs = {
  friendId: Scalars['ID'];
};


export type MutationDeleteFriendArgs = {
  friendId: Scalars['ID'];
};


export type MutationCreateChannelArgs = {
  channel?: Maybe<ChannelInput>;
};


export type MutationUpdateChannelArgs = {
  channel?: Maybe<ChannelInput>;
};


export type MutationDeleteChannelArgs = {
  channelId: Scalars['ID'];
};


export type MutationCreateMessageArgs = {
  message: Scalars['String'];
  channelId: Scalars['String'];
};


export type MutationSetOnlineStatusArgs = {
  isOnline?: Maybe<Scalars['Boolean']>;
};


export type MutationChangeEmailPasswordArgs = {
  password: Scalars['String'];
  newPassword: Scalars['String'];
};


export type MutationCreateGalleryArgs = {
  photoURL: Scalars['String'];
};


export type MutationUpdateGalleryArgs = {
  galleryId: Scalars['ID'];
  photoURL: Scalars['String'];
};


export type MutationDeleteGalleryArgs = {
  galleryId: Scalars['ID'];
};


export type MutationSingleUploadArgs = {
  file: Scalars['Upload'];
  dir?: Maybe<Scalars['String']>;
};


export type MutationCreateReactionArgs = {
  messageId: Scalars['ID'];
  type: Scalars['String'];
};


export type MutationDeleteReactionArgs = {
  reactionId: Scalars['ID'];
};

export type Notification = {
  __typename?: 'Notification';
  id: Scalars['ID'];
  token?: Maybe<Scalars['String']>;
  device?: Maybe<Scalars['String']>;
  os?: Maybe<Scalars['String']>;
  createdAt?: Maybe<Scalars['DateTime']>;
  updatedAt?: Maybe<Scalars['DateTime']>;
};

export type NotificationCreateInput = {
  token: Scalars['String'];
  device?: Maybe<Scalars['String']>;
  os?: Maybe<Scalars['String']>;
};

/** Information about pagination in a connection. */
export type PageInfo = {
  __typename?: 'PageInfo';
  startCursor?: Maybe<Scalars['String']>;
  endCursor?: Maybe<Scalars['String']>;
  hasNextPage?: Maybe<Scalars['Boolean']>;
  hasPreviousPage?: Maybe<Scalars['Boolean']>;
};

export type Photo = {
  __typename?: 'Photo';
  id: Scalars['String'];
  thumbURL?: Maybe<Scalars['String']>;
  photoURL?: Maybe<Scalars['String']>;
  createdAt?: Maybe<Scalars['DateTime']>;
  updatedAt?: Maybe<Scalars['DateTime']>;
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export type Query = {
  __typename?: 'Query';
  /**
   * If filter is true, it will filter user with email, nickname or name.
   * You can add pagination with first and after args.
   */
  users?: Maybe<UsersConnection>;
  user?: Maybe<User>;
  me?: Maybe<User>;
  messages: Array<Message>;
  channels: Array<Channel>;
  friends: Array<User>;
  galleries: Array<Gallery>;
};


export type QueryUsersArgs = {
  user?: Maybe<UserQueryInput>;
  includeUser?: Maybe<Scalars['Boolean']>;
  filter?: Maybe<Scalars['Boolean']>;
  first?: Maybe<Scalars['Int']>;
  last?: Maybe<Scalars['Int']>;
  before?: Maybe<Scalars['String']>;
  after?: Maybe<Scalars['String']>;
};


export type QueryUserArgs = {
  id: Scalars['ID'];
};


export type QueryGalleriesArgs = {
  userId: Scalars['String'];
};

export type Reaction = {
  __typename?: 'Reaction';
  id: Scalars['ID'];
  type?: Maybe<Scalars['String']>;
  createdAt?: Maybe<Scalars['DateTime']>;
  updatedAt?: Maybe<Scalars['DateTime']>;
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export type Reply = {
  __typename?: 'Reply';
  id: Scalars['String'];
  sender?: Maybe<User>;
  type?: Maybe<Scalars['String']>;
  text?: Maybe<Scalars['String']>;
  filePath?: Maybe<Scalars['String']>;
  replies?: Maybe<Array<Maybe<Reply>>>;
  createdAt?: Maybe<Scalars['DateTime']>;
  updatedAt?: Maybe<Scalars['DateTime']>;
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export type SocialUserInput = {
  socialId: Scalars['String'];
  authType: AuthType;
  email?: Maybe<Scalars['String']>;
  photoURL?: Maybe<Scalars['String']>;
  thumbURL?: Maybe<Scalars['String']>;
  name?: Maybe<Scalars['String']>;
  nickname?: Maybe<Scalars['String']>;
  birthday?: Maybe<Scalars['Date']>;
  gender?: Maybe<Gender>;
  phone?: Maybe<Scalars['String']>;
};

export type Subscription = {
  __typename?: 'Subscription';
  userSignedIn?: Maybe<User>;
  userUpdated?: Maybe<User>;
  friendChanged?: Maybe<FriendSub>;
};


export type SubscriptionUserSignedInArgs = {
  userId: Scalars['ID'];
};


export type SubscriptionUserUpdatedArgs = {
  userId: Scalars['ID'];
};


export type SubscriptionFriendChangedArgs = {
  userId: Scalars['ID'];
};


export type User = {
  __typename?: 'User';
  id: Scalars['ID'];
  email?: Maybe<Scalars['String']>;
  name?: Maybe<Scalars['String']>;
  nickname?: Maybe<Scalars['String']>;
  thumbURL?: Maybe<Scalars['String']>;
  photoURL?: Maybe<Scalars['String']>;
  birthday?: Maybe<Scalars['Date']>;
  gender?: Maybe<Gender>;
  socialId?: Maybe<Scalars['String']>;
  authType?: Maybe<AuthType>;
  phone?: Maybe<Scalars['String']>;
  verified?: Maybe<Scalars['Boolean']>;
  statusMessage?: Maybe<Scalars['String']>;
  isOnline?: Maybe<Scalars['Boolean']>;
  lastSignedIn?: Maybe<Scalars['DateTime']>;
  notifications?: Maybe<Array<Maybe<Notification>>>;
  createdAt?: Maybe<Scalars['DateTime']>;
  updatedAt?: Maybe<Scalars['DateTime']>;
  deletedAt?: Maybe<Scalars['DateTime']>;
};

export type UserEdge = {
  __typename?: 'UserEdge';
  node?: Maybe<User>;
  cursor?: Maybe<Scalars['String']>;
};

export type UserInput = {
  email: Scalars['String'];
  password: Scalars['String'];
  name?: Maybe<Scalars['String']>;
  nickname?: Maybe<Scalars['String']>;
  birthday?: Maybe<Scalars['Date']>;
  gender?: Maybe<Gender>;
  phone?: Maybe<Scalars['String']>;
  statusMessage?: Maybe<Scalars['String']>;
};

export enum UserModeType {
  Default = 'DEFAULT',
  Hidden = 'HIDDEN',
  Block = 'BLOCK'
}

export type UserProfileInput = {
  name?: Maybe<Scalars['String']>;
  nickname?: Maybe<Scalars['String']>;
  birthday?: Maybe<Scalars['Date']>;
  gender?: Maybe<Gender>;
  phone?: Maybe<Scalars['String']>;
  thumbURL?: Maybe<Scalars['String']>;
  photoURL?: Maybe<Scalars['String']>;
  statusMessage?: Maybe<Scalars['String']>;
};

export type UserQueryInput = {
  email?: Maybe<Scalars['String']>;
  name?: Maybe<Scalars['String']>;
  nickname?: Maybe<Scalars['String']>;
  birthday?: Maybe<Scalars['Date']>;
  gender?: Maybe<Gender>;
  phone?: Maybe<Scalars['String']>;
};

/**
 * Simple wrapper around our list of launches that contains a cursor to the
 * last item in the list. Pass this cursor to the launches query to fetch results
 * after these.
 */
export type UsersConnection = {
  __typename?: 'UsersConnection';
  totalCount?: Maybe<Scalars['Int']>;
  edges?: Maybe<Array<Maybe<UserEdge>>>;
  pageInfo?: Maybe<PageInfo>;
};
