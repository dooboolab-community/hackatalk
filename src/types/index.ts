import { StyleProp, TextStyle } from 'react-native';

import { SFC } from 'react';

export enum MessageType {
  Message,
  Photo,
  File,
}

export enum AuthType {
  EMAIL,
  FACEBOOK,
  GOOGLE,
  APPLE
}

export interface User {
  id: string;
  email?: string;
  nickname?: string;
  birthday?: Date;
  name?: string;
  thumbURL?: string;
  photoURL?: string;
  statusMessage?: string;
  isOnline?: boolean;
  socialId?: string;
  lastSignedIn?: string;
  authType?: AuthType;
  verified?: boolean;
  created?: string;
  updated?: string;
}

export interface AuthPayload {
  token: string;
  user: User;
}

export interface Friend extends User {
  isFriend?: boolean;
  friendSince: Date;
}

export interface MessageCommon<T extends MessageType = MessageType.Message> {
  id: string;
  sender: User;
  messageType?: T;
  created?: string;
  updated?: string;
}

interface Message extends MessageCommon<MessageType.Message> {
  message: string;
}

interface Photo extends MessageCommon<MessageType.Photo> {
  photo: string;
}

interface File extends MessageCommon<MessageType.File> {
  file: string;
}

export type MessageProps = Message | Photo | File;

export interface ChannelType {
  id: string;
  lastMessage: MessageProps;
  lastMessageCnt: number;
  messages?: MessageProps[];
  users?: User[];
  created?: Date;
  updated?: Date;
}

// export type Friend = Omit<User, 'uid'>;

interface IconProps {
  style?: StyleProp<TextStyle>;
  width?: number | string;
  height?: number | string;
  children?: never;
}

export type IconType = SFC<IconProps>;

export enum ThemeType {
  LIGHT = 'LIGHT',
  DARK = 'DARK',
}
