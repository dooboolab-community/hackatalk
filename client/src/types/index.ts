import { StyleProp, TextStyle } from 'react-native';

import { SFC } from 'react';
import { User } from './graphql';

export enum MessageType {
  Message,
  Photo,
  File,
}

export interface Friend extends User {
  isFriend?: boolean;
  friendSince: Date;
}

interface IconProps {
  style?: StyleProp<TextStyle>;
  width?: number | string;
  height?: number | string;
  children?: never;
}

export type IconType = SFC<IconProps>;

export enum ThemeType {
  LIGHT = 'LIGHT',
  DARK = 'DARK',
}
