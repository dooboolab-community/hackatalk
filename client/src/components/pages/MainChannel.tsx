import type {
  ChannelsQuery,
  ChannelsQueryResponse,
  ChannelsQueryVariables,
} from '../../__generated__/ChannelsQuery.graphql';
import {FlatList, Platform, TouchableOpacity, View} from 'react-native';
import React, {FC, Suspense, useMemo, useState} from 'react';
import {graphql, useLazyLoadQuery, usePaginationFragment} from 'react-relay';
import useOrientation, {Orientation} from '../../hooks/useOrientation';

import {AdMobBanner} from 'expo-ads-admob';
import {Channel} from '../../types/graphql';
import ChannelListItem from '../uis/ChannelListItem';
import CustomLoadingIndicator from '../uis/CustomLoadingIndicator';
import EmptyListItem from '../uis/EmptyListItem';
import type {MainChannelComponent_channel$key} from '../../__generated__/MainChannelComponent_channel.graphql';
import {MainStackNavigationProps} from '../navigations/MainStackNavigator';
import {MaterialTopTabNavigationProps} from '../navigations/MainTabNavigator';
import {SvgPlus} from '../../utils/Icons';
import {channelsQuery} from '../../relay/queries/Channel';
import {getString} from '../../../STRINGS';
import styled from '@emotion/native';
import {useNavigation} from '@react-navigation/native';
import {useTheme} from 'dooboo-ui';

const Container = styled.View`
  flex: 1;
  background: ${({theme}) => theme.background};
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Fab = styled.View`
  width: 56px;
  height: 56px;
  border-radius: 28px;
  justify-content: center;
  align-items: center;
  background: ${({theme}) => theme.fab};
`;

const ITEM_CNT = 20;

const channelsPaginationFragment = graphql`
  fragment MainChannelComponent_channel on Query
  @argumentDefinitions(
    first: {type: "Int!"}
    after: {type: "String"}
    withMessage: {type: "Boolean"}
  )
  @refetchable(queryName: "Channels") {
    channels(first: $first, after: $after, withMessage: $withMessage)
      @connection(
        key: "MainChannelComponent_channels"
        filters: ["withMessage"]
      ) {
      edges {
        cursor
        node {
          id
          channelType
          name
          memberships(excludeMe: true) {
            user {
              name
              nickname
              thumbURL
              photoURL
            }
          }
          lastMessage {
            id
            messageType
            text
            imageUrls
            fileUrls
            createdAt
          }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;

interface ChannelProps {
  channel: MainChannelComponent_channel$key;
  searchArgs: ChannelsQueryVariables;
}

const ChannelsFragment: FC<ChannelProps> = ({channel, searchArgs}) => {
  const {data, loadNext, isLoadingNext, refetch} = usePaginationFragment<
    ChannelsQuery,
    MainChannelComponent_channel$key
  >(channelsPaginationFragment, channel);

  const [bannerError, setBannerError] = useState<boolean>(false);
  const orientation = useOrientation();
  const navigation = useNavigation<MainStackNavigationProps<'MainTab'>>();

  const onEndReached = (): void => {
    loadNext(ITEM_CNT);
  };

  const renderItem = ({
    item,
    index,
  }: {
    item: {node: Channel; cursor: string};
    index: number;
  }): React.ReactElement | null => {
    if (!item.node?.memberships || !item.node?.memberships.length)
      return <View key={index} />;

    return (
      <ChannelListItem
        testID={`list-item-${index}`}
        key={index.toString()}
        item={item.node}
        onPress={(): void => {
          navigation.navigate('Message', {
            channelId: item.node.id,
          });
        }}
      />
    );
  };

  const channels = useMemo(() => {
    return data?.channels?.edges ?? [];
  }, [data?.channels?.edges]);

  return (
    <FlatList
      scrollIndicatorInsets={{right: 1}}
      style={{
        alignSelf: 'stretch',
      }}
      contentContainerStyle={
        channels.length === 0
          ? {
              flex: 1,
              alignSelf: 'stretch',
              alignItems: 'center',
              justifyContent: 'center',
            }
          : null
      }
      keyExtractor={(_, index): string => index.toString()}
      // @ts-ignore
      data={channels}
      renderItem={renderItem}
      ListHeaderComponent={
        !bannerError && orientation === Orientation.PORTRAIT
          ? Platform.select({
              android: (
                <AdMobBanner
                  bannerSize={'smartBannerPortrait'}
                  // adUnitID="ca-app-pub-3940256099942544/6300978111"
                  adUnitID="ca-app-pub-7837089095803162/8109702961"
                  onDidFailToReceiveAdWithError={() => setBannerError(true)}
                />
              ),
              ios: (
                <AdMobBanner
                  bannerSize={'smartBannerPortrait'}
                  // adUnitID="ca-app-pub-3940256099942544/2934735716"
                  adUnitID="ca-app-pub-7837089095803162/4326063134"
                  onDidFailToReceiveAdWithError={() => setBannerError(true)}
                />
              ),
            })
          : null
      }
      ListEmptyComponent={
        <EmptyListItem>{getString('NO_CHANNELLIST')}</EmptyListItem>
      }
      ListFooterComponent={<View style={{height: 60}} />}
      refreshing={isLoadingNext}
      onRefresh={() => {
        refetch(searchArgs, {fetchPolicy: 'network-only'});
      }}
      onEndReachedThreshold={0.1}
      onEndReached={onEndReached}
    />
  );
};

interface ContentProps {
  searchArgs: ChannelsQueryVariables;
}

const ContentContainer: FC<ContentProps> = ({searchArgs}) => {
  const data: ChannelsQueryResponse = useLazyLoadQuery<ChannelsQuery>(
    channelsQuery,
    searchArgs,
    {fetchPolicy: 'store-or-network'},
  );

  return <ChannelsFragment channel={data} searchArgs={searchArgs} />;
};

const Screen: FC = () => {
  const {theme} = useTheme();
  const navigation = useNavigation<MaterialTopTabNavigationProps<'Channel'>>();

  const searchArgs: ChannelsQueryVariables = {
    first: ITEM_CNT,
    withMessage: true,
  };

  return (
    <Container>
      <Suspense fallback={<CustomLoadingIndicator />}>
        <ContentContainer searchArgs={searchArgs} />
      </Suspense>
      <TouchableOpacity
        activeOpacity={0.65}
        style={{
          position: 'absolute',
          right: 20,
          bottom: 44,
        }}
        onPress={(): void => navigation.navigate('ChannelCreate')}>
        <Fab>
          <SvgPlus fill={theme.background} />
        </Fab>
      </TouchableOpacity>
    </Container>
  );
};

export default Screen;
