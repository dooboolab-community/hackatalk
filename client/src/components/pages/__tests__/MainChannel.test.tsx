import 'react-native';

import * as React from 'react';

import {MockPayloadGenerator, createMockEnvironment} from 'relay-test-utils';
import {
  createMockNavigation,
  createTestElement,
} from '../../../../test/testUtils';
import {fireEvent, render} from '@testing-library/react-native';

import {Channel} from '../../../types/graphql';
import MainChannel from '../MainChannel';
import ReactNavigation from '@react-navigation/core';

const mockNavigation = createMockNavigation();

jest.mock('@react-navigation/core', () => ({
  ...jest.requireActual<typeof ReactNavigation>('@react-navigation/core'),
  useNavigation: () => mockNavigation,
}));

const mockEnvironment = createMockEnvironment();

mockEnvironment.mock.queueOperationResolver((operation) =>
  MockPayloadGenerator.generate(operation, {
    Channel: (_, generateId): Channel => ({
      id: `test-channel-${generateId()}`,
      name: 'HackaTalk',
      channelType: 'private',
      lastMessage: {
        id: 'test-message-3848',
        messageType: 'text',
        createdAt: '2021-03-19T05:13:22.932Z',
      },
      messages: {
        edges: [],
        pageInfo: {
          hasNextPage: false,
          hasPreviousPage: false,
        },
      },
    }),
  }),
);

describe('[Channel] screen', () => {
  it('renders without crashing', () => {
    const component = createTestElement(<MainChannel />, {
      environment: mockEnvironment,
    });

    const screen = render(component);
    const json = screen.toJSON();

    expect(json).toBeTruthy();
    expect(json).toMatchSnapshot();
  });

  describe('interactions', () => {
    it('should simulate onPress', async () => {
      const component = createTestElement(<MainChannel />, {
        environment: mockEnvironment,
      });

      const screen = render(component);
      const btn = await screen.findByTestId('list-item-0');

      fireEvent.press(btn);

      // TODO: Test what happens after pressing the button.
    });
  });
});
