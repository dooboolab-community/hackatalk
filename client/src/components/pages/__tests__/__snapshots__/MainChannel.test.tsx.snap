// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Channel] screen renders without crashing 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#000",
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <RCTScrollView
      ListEmptyComponent={
        <Shared
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "flex": 1,
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          Start a Chat
        </Shared>
      }
      ListFooterComponent={
        <View
          style={
            Object {
              "height": 60,
            }
          }
        />
      }
      ListHeaderComponent={
        <AdMobBanner
          adUnitID="ca-app-pub-7837089095803162/4326063134"
          bannerSize="smartBannerPortrait"
          onDidFailToReceiveAdWithError={[Function]}
        />
      }
      contentContainerStyle={null}
      data={
        Array [
          Object {
            "cursor": "<mock-value-for-field-\\"cursor\\">",
            "node": Object {
              "__typename": "Channel",
              "channelType": "private",
              "id": "test-channel-1",
              "lastMessage": Object {
                "createdAt": "2021-03-19T05:13:22.932Z",
                "fileUrls": "<mock-value-for-field-\\"fileUrls\\">",
                "id": "test-message-3848",
                "imageUrls": "<mock-value-for-field-\\"imageUrls\\">",
                "messageType": "text",
                "text": "<mock-value-for-field-\\"text\\">",
              },
              "memberships": Array [
                Object {
                  "user": Object {
                    "id": "<User-mock-id-2>",
                    "name": "<mock-value-for-field-\\"name\\">",
                    "nickname": "<mock-value-for-field-\\"nickname\\">",
                    "photoURL": "<mock-value-for-field-\\"photoURL\\">",
                    "thumbURL": "<mock-value-for-field-\\"thumbURL\\">",
                  },
                },
              ],
              "name": "HackaTalk",
            },
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={0.1}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      scrollIndicatorInsets={
        Object {
          "right": 1,
        }
      }
      stickyHeaderIndices={Array []}
      style={
        Object {
          "alignSelf": "stretch",
        }
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <RCTRefreshControl />
      <View>
        <View
          onLayout={[Function]}
        >
          <View>
            <ViewManagerAdapter_ExpoAdsAdMobBannerView
              onDidFailToReceiveAdWithError={[Function]}
              onSizeChange={[Function]}
              proxiedProperties={
                Object {
                  "adUnitID": "ca-app-pub-7837089095803162/4326063134",
                  "additionalRequestParams": Object {
                    "npa": "1",
                  },
                  "bannerSize": "smartBannerPortrait",
                }
              }
              style={Object {}}
            />
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "justifyContent": "center",
                "width": "100%",
              }
            }
          >
            <View
              accessibilityLabel="Go to chat"
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#151C22",
                      "borderBottomColor": "#515151",
                      "borderBottomWidth": 0.3,
                      "borderLeftColor": "#515151",
                      "borderRightColor": "#515151",
                      "borderTopColor": "#515151",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "minHeight": 92,
                      "paddingBottom": 8,
                      "paddingLeft": 0,
                      "paddingRight": 0,
                      "paddingTop": 8,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "marginHorizontal": 15,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "height": 50,
                        "justifyContent": "center",
                        "width": 50,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flexDirection": "row",
                          "flexWrap": "wrap",
                          "height": 44,
                          "width": 44,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flexDirection": "row",
                          }
                        }
                      >
                        <Image
                          source={
                            Object {
                              "uri": "<mock-value-for-field-\\"thumbURL\\">",
                            }
                          }
                          style={
                            Object {
                              "borderRadius": 24,
                              "height": 48,
                              "width": 48,
                            }
                          }
                        />
                      </View>
                    </View>
                  </View>
                  <View />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "flexBasis": 0,
                        "flexDirection": "column",
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "paddingRight": 20,
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                        },
                        undefined,
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "flexDirection": "row",
                            "justifyContent": "flex-start",
                          },
                          undefined,
                        ]
                      }
                    >
                      <Text
                        numberOfLines={2}
                        style={
                          Array [
                            Object {
                              "color": "white",
                              "fontSize": 14,
                            },
                            Object {
                              "fontWeight": "bold",
                            },
                          ]
                        }
                      >
                        &lt;mock-value-for-field-"nickname"&gt;
                      </Text>
                    </View>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                          "marginTop": 8,
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      lastMessageCnt={0}
                      numberOfLines={2}
                      style={
                        Array [
                          Object {
                            "color": "white",
                            "fontSize": 12,
                            "maxWidth": 200,
                          },
                          undefined,
                        ]
                      }
                    >
                      &lt;mock-value-for-field-"text"&gt;
                    </Text>
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "white",
                            "fontSize": 12,
                            "textAlign": "right",
                          },
                          undefined,
                        ]
                      }
                    >
                      in 42 minutes
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Object {
                "height": 60,
              }
            }
          />
        </View>
      </View>
    </RCTScrollView>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 44,
          "opacity": 1,
          "position": "absolute",
          "right": 20,
        }
      }
      testID="channel-create-fab"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#8BC9FF",
              "borderBottomLeftRadius": 28,
              "borderBottomRightRadius": 28,
              "borderTopLeftRadius": 28,
              "borderTopRightRadius": 28,
              "height": 56,
              "justifyContent": "center",
              "width": 56,
            },
            undefined,
          ]
        }
      >
        <SvgMock
          fill="#fff"
        />
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
